server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -277,61 +276,101 @@ private boolean skipRepository(String repositoryName) {
             } else {
                 assert fromSortValuePredicates.isMatchAll() : "filtering is not supported in non-verbose mode";
                 assert slmPolicyPredicate == SlmPolicyPredicate.MATCH_ALL_POLICIES : "filtering is not supported in non-verbose mode";
 
                 addSimpleSnapshotInfos(
-                    toResolve,
-                    repo,
+                    getSnapshotIdIterator(repositoryName, repositoryData),
+                    repositoryName,
                     repositoryData,
-                    snapshotsInProgress.forRepo(repo).stream().map(entry -> SnapshotInfo.inProgress(entry).basic()).toList()
+                    snapshotsInProgress.forRepo(repositoryName).stream().map(entry -> SnapshotInfo.inProgress(entry).basic()).toList()
                 );
                 listener.onResponse(null);
             }
         }
 
-        private void loadSnapshotInfos(String repositoryName, Collection<SnapshotId> snapshotIds, ActionListener<Void> listener) {
+        /**
+         * Check that the repository contains every <i>required</i> name according to {@link #snapshotNamePredicate}.
+         *
+         * @throws SnapshotMissingException if one or more required names are missing.
+         */
+        private void ensureRequiredNamesPresent(String repositoryName, @Nullable RepositoryData repositoryData) {
+            if (snapshotNamePredicate.requiredNames().isEmpty()) {
+                return;
+            }
+
+            final var unmatchedRequiredNames = new HashSet<>(snapshotNamePredicate.requiredNames());
+            for (final var snapshotInProgress : snapshotsInProgress.forRepo(repositoryName)) {
+                unmatchedRequiredNames.remove(snapshotInProgress.snapshot().getSnapshotId().getName());
+            }
+            if (unmatchedRequiredNames.isEmpty()) {
+                return;
+            }
+            if (repositoryData != null) {
+                for (final var snapshotId : repositoryData.getSnapshotIds()) {
+                    unmatchedRequiredNames.remove(snapshotId.getName());
+                }
+                if (unmatchedRequiredNames.isEmpty()) {
+                    return;
+                }
+            }
+            throw new SnapshotMissingException(repositoryName, unmatchedRequiredNames.iterator().next());
+        }
+
+        /**
+         * @return an iterator over the snapshot IDs in the given repository which match {@link #snapshotNamePredicate}.
+         */
+        private Iterator<SnapshotId> getSnapshotIdIterator(String repositoryName, @Nullable RepositoryData repositoryData) {
+
+            // now iterate through the snapshots again, returning matching IDs (or null)
+            final Set<SnapshotId> matchingInProgressSnapshots = new HashSet<>();
+            return Iterators.concat(
+                // matching in-progress snapshots first
+                Iterators.filter(
+                    Iterators.map(
```

Comment:
Not related to this PR. Why you prefer static iter constructors over method chaining like stream api. Reading inside-out not fun at all :) 

Isnt it easier to read? Does not looks to complex to extend current iterators.

```java
        snapshotsInProgress.forRepo(repositoryName)
            .iterator()
            .map(snapshotInProgress -> snapshotInProgress.snapshot().getSnapshotId())
            .filter(snapshotId -> {
                if (snapshotNamePredicate.test(snapshotId.getName(), true)) {
                    matchingInProgressSnapshots.add(snapshotId);
                    return true;
                } else {
                    return false;
                }
            })
            .concat(
                () -> repositoryData == null
                    ? Collections.emptyIterator()
                    : repositoryData.getSnapshotIds()
                        .iterator()
                        .filter(
                            snapshotId -> matchingInProgressSnapshots.contains(snapshotId) == false
                                && snapshotNamePredicate.test(snapshotId.getName(), false)
                                && matchesPredicates(snapshotId, repositoryData)
                        ));
```

Reply:
The stream APIs are very much more expressive than simple iterators, which is all very nice, but it turns out that this means they're outrageously expensive at runtime as a result.

Reply:
I didnt mean to use stream, but rather allow iterator have non-static method to create another iter using map or filter. It might call current static method that we have. So we can chain them, I dont think it cost much.

```java
    class Iterator <T>{
        public <U> Iterator<U> map(Function<T, U> mapFn) {
            return Iterators.map(this, mapFn);
        }
        
        public Iterator<T> filter(Predicate<T> filterFn) {
            return Iterators.filter(this, filterFn);
        }
        
        public static void main(String[] args) {
            var iter = new Iterator<String>();
            iter.map(String::length)
                .filter(l -> l>0);
        }
    }
```

Reply:
Ah right yes I'd love to do that but this is `java.util.Iterator<T>`, it's in the JDK, so not something to which we can add methods ourselves. We could have our own `Iterator` whose interface we could extend, but then we'd end up having to add layers of wrapping to adapt it into the JDK one and back again and it'd end up being fairly messy in practice.